client/src/components/Tool/ToolForm.vue

```diff
@@ -0,0 +1,300 @@
+                const extra_user_preferences = JSON.parse(Galaxy.user.attributes.preferences.extra_user_preferences);
+                const keyCached = "use_cached_job|use_cached_job_checkbox";
+                const hasCachedJobs = keyCached in extra_user_preferences;
+                return hasCachedJobs ? extra_user_preferences[keyCached] : false;
+            } else {
+                return false;
+            }
+        },
+    },
+    methods: {
+        onValidation(validationInternal) {
+            this.validationInternal = validationInternal;
+        },
+        onChange(newData) {
+            this.formData = newData;
+            this.onUpdate();
+        },
+        onUpdate() {
+            updateTool(this.formConfig.id, this.currentVersion, this.formData).then((data) => {
+                this.formConfig = data;
+            });
+        },
+        onChangeVersion(newVersion) {
+            this.requestTool(newVersion);
+        },
+        onChangeEmail(email) {
+            this.email = email;
+        },
+        onChangeRemap(remap) {
+            this.remap = remap ? this.job_id : null;
+        },
+        onChangeReuse(reuse) {
+            this.reuse = reuse;
+        },
+        onUpdateFavorites(user, newFavorites) {
+            user.preferences["favorites"] = newFavorites;
+        },
+        requestTool(newVersion) {
+            this.currentVersion = newVersion || this.currentVersion;
+            getTool(this.id, this.currentVersion, this.job_id).then((data) => {
+                this.formConfig = data;
+                this.showLoading = false;
+                this.showForm = true;
+                if (newVersion) {
+                    this.messageVariant = "success";
+                    this.messageText = `Now you are using '${data.name}' version ${data.version}, id '${data.id}'.`;
+                }
+            });
+        },
+        onExecute() {
+            if (this.validationInternal) {
+                this.validationScrollTo = this.validationInternal.slice();
+                return;
+            }
+            this.showExecuting = true;
+            const options = this.formData;
+            const Galaxy = getGalaxyInstance();
+            const history_id = Galaxy.currHistoryPanel && Galaxy.currHistoryPanel.model.id;
+            const jobDef = {
+                history_id: history_id,
+                tool_id: this.formConfig.id,
+                tool_version: this.formConfig.version,
+                inputs: this.formData,
+            };
+            if (this.formConfig.action !== `${getAppRoot()}tool_runner/index`) {
+                send(options, jobDef);
+                return;
+            }
+            Galaxy.emit.debug("tool-form::submit()", "Validation complete.", jobDef);
+            submitJob(jobDef).then(
+                (jobResponse) => {
+                    if (Galaxy.currHistoryPanel) {
+                        Galaxy.currHistoryPanel.refreshContents();
```

Comment:
Should these be events that are handled externally ?

Reply:
I would argue yes. We should be emitting events and the outside should decide how to deal with them.

Reply:
Yes, please see my comment above. Also currently the `ToolForm` component is directly loaded from the `Analysis` router which has not been converted to Vue yet. These two insertions which rely on the global Galaxy object are only required while the current history panel is on by default.